src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,430 @@
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CompleteStatementCommandHandler()
+        {
+        }
+
+        public string DisplayName => CSharpEditorResources.Complete_statement_on_semicolon;
+
+        public void ExecuteCommand(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            // Determine where semicolon should be placed and move caret to location
+            BeforeExecuteCommand(args, executionContext);
+
+            // Insert the semicolon using next command handler
+            nextCommandHandler();
+        }
+
+        private void BeforeExecuteCommand(TypeCharCommandArgs args, CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';')
+            {
+                return;
+            }
+
+            var caretOpt = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caretOpt.HasValue)
+            {
+                return;
+            }
+
+            var caret = caretOpt.Value;
+
+            var document = caret.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return;
+            }
+
+            var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+            var caretPosition = caret.Position;
+
+            var token = root.FindToken(caretPosition);
+
+            var currentNode = token.Parent;
+
+            // If cursor is right before an opening delimiter, start with node outside of delimiters since analysis 
+            // starting with a node containing delimiters assumes the caret is placed inside those delimiters. 
+            // This covers cases like `obj.ToString$()`, where `token` references `(` but the caret isn't actually 
+            // inside the argument list.
+            if (token.IsKind(SyntaxKind.OpenBraceToken, SyntaxKind.OpenBracketToken, SyntaxKind.OpenParenToken)
+                && token.Span.Start >= caretPosition)
+            {
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode == null)
+            {
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (!LooksLikeNodeInArgumentList(currentNode, caret, syntaxFacts))
+            {
+                return;
+            }
+
+            // verify all delimeters exist until you reach statement syntax that requires a semicolon
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                if (RequiredDelimiterIsMissing(currentNode))
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
```

Comment:
it would be good to have an example here as well, like you hve below with StatementClosingDelimiterIsMissing

Reply:
Added "// Example: missing final `)` in `obj.Method($`"